tests/test_solving.py

```diff
@@ -230,3 +230,55 @@ def dep3(
 
     container = Container()
     container.solve(Dependant(dep3, scope=None), scopes=[None])
+
+
+def test_infer_scope_1() -> None:
+    # This dep must be inferred into the "app" scope
+    # because that's the only scope we enter
+    # Otherwise we'd get an error
+
+    def dep() -> Iterator[None]:
+        yield None
+
+    container = Container()
+    solved = container.solve(Dependant(dep), scopes=["app"])
+    with container.enter_scope("app") as state:
+        container.execute_sync(solved, SyncExecutor(), state=state)
+
+
+def test_infer_scope_2() -> None:
+    # Since None is a known scope, we don't change it
+
+    def dep() -> Iterator[None]:
+        yield None
+
+    container = Container()
+    solved = container.solve(Dependant(dep), scopes=[None])
+    with container.enter_scope(None) as state:
+        container.execute_sync(solved, SyncExecutor(), state=state)
+
+
+def test_infer_scope_3() -> None:
+    def db_connection() -> None:
+        ...
+
+    DataBase = Annotated[None, Marker(db_connection, scope="app")]
+
+    def query_param() -> None:
+        ...
+
+    QueryParam = Annotated[None, Marker(query_param, scope="request")]
+
+    def user_function(db: DataBase, param: QueryParam) -> None:
+        ...
+
+    def endpoint(func: Annotated[None, Marker(user_function)]) -> None:
+        ...
```

Comment:
One pretty cool outcome of this is that if you define an endpoint that does not depend on an HTTP request (e.g. serving static content) it will automatically get cached:

```python
def endpoint() -> str:
    return "foo"
```

(in this case an many others it would be less efficient to cache vs. call the function, but that's besides the point, caching has a fixed cost)